Classes/BEMSimpleLineGraphView.m

```diff
@@ -1264,55 +1257,82 @@ - (void)handleGestureAction:(UIGestureRecognizer *)recognizer {
+        if (numberOfPoints == 0) return; //something's very wrong
     }
+    closestDot.alpha = 0.8f;
 
-    if (closestDot.tag >= DotFirstTag100  && [closestDot isMemberOfClass:[BEMCircle class]]) {
-        if ([self.delegate respondsToSelector:@selector(lineGraph:didTouchGraphWithClosestIndex:)] && self.enableTouchReport == YES) {
-            [self.delegate lineGraph:self didTouchGraphWithClosestIndex:((NSUInteger)closestDot.tag - DotFirstTag100)];
-
+    if (recognizer.state != UIGestureRecognizerStateEnded) {
+        //ON START OR MOVE
+        if (self.enablePopUpReport == YES  && self.alwaysDisplayPopUpLabels == NO) {
+            if (!self.customPopUpView && [self.delegate respondsToSelector:@selector(popUpViewForLineGraph:)] ) {
+                self.customPopUpView = [self.delegate popUpViewForLineGraph:self];
+            }
+            if (self.customPopUpView) {
+                [self addSubview:self.customPopUpView];
+                [self adjustXLocForLabel:self.customPopUpView avoidingDot:closestDot.frame];
+                [self adjustYLocForLabel:self.customPopUpView avoidingDot:closestDot.frame andNeighbors:CGRectZero and:CGRectZero];
+                if ([self.delegate respondsToSelector:@selector(lineGraph:modifyPopupView:forIndex:)]) {
+                    self.customPopUpView.alpha = 1.0f;
+                    [self.delegate lineGraph:self modifyPopupView:self.customPopUpView forIndex:index];
+                } else {
+                    [UIView animateWithDuration:0.2f delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^{
+                        self.customPopUpView.alpha = 1.0f;
+                    } completion:nil];
+                }
+            } else {
+                if (!self.popUpLabel) {
+                    self.popUpLabel = [[UILabel alloc] initWithFrame:CGRectZero];
+                }
+                [self addSubview: self.popUpLabel ];
+                self.popUpLabel = [self configureLabel:self.popUpLabel forPoint:closestDot];
+                [self adjustXLocForLabel:self.popUpLabel avoidingDot:closestDot.frame];
+                [self adjustYLocForLabel:self.popUpLabel avoidingDot:closestDot.frame andNeighbors:CGRectZero and:CGRectZero];
+                [UIView animateWithDuration:0.2f delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^{
+                    self.popUpLabel.alpha = 1.0f;
+                } completion:nil];
+            }
         }
-    }
-
-    // ON RELEASE
-    if (recognizer.state == UIGestureRecognizerStateEnded) {
-        if ([self.delegate respondsToSelector:@selector(lineGraph:didReleaseTouchFromGraphWithClosestIndex:)]) {
-            [self.delegate lineGraph:self didReleaseTouchFromGraphWithClosestIndex:(closestDot.tag - DotFirstTag100)];
 
+        if (self.enableTouchReport && [self.delegate respondsToSelector:@selector(lineGraph:didTouchGraphWithClosestIndex:)]) {
+            [self.delegate lineGraph:self didTouchGraphWithClosestIndex:index];
```

Comment:
Hmm, I don't think so; circleDots aren't created when there's no value, so closestDotfromTouchInputLine can't see them.  I tried it in TestBed and it didn't show null values.